test/task.cc

```diff
@@ -392,3 +392,102 @@ TEST(Task, Failure) {
 
   EXPECT_THROW(*e(), std::exception_ptr);
 }
+
+TEST(Task, Inheritance) {
+  struct Base {
+    virtual Task::Of<int> GetTask() = 0;
+  };
+
+  struct Sync : public Base {
+    Task::Of<int> GetTask() override {
+      return Task::Success(10);
+    }
+  };
+
+  struct Async : public Base {
+    Task::Of<int> GetTask() override {
+      return []() {
+        return Just(20);
+      };
+    }
+  };
+
+  struct Failure : public Base {
+    Task::Of<int> GetTask() override {
+      return Task::Failure("error");
+    }
+  };
+
+  auto f = []() -> Task::Of<void> {
+    return []() {
+      return Just();
+    };
+  };
+
+  auto sync = [&]() {
+    return f()
+        | Sync().GetTask();
+  };
+
+  auto async = [&]() {
+    return f()
+        | Async().GetTask();
+  };
+
+  auto failure = [&]() {
+    return f()
+        | Failure().GetTask();
+  };
+
+  EXPECT_EQ(*sync(), 10);
+  EXPECT_EQ(*async(), 20);
+  EXPECT_THROW(*failure(), std::exception_ptr);
+}
+
+TEST(Task, MoveableSuccess) {
+  auto e = []() -> Task::Of<std::unique_ptr<int>> {
+    return Task::Success(std::make_unique<int>(10));
+  };
+
+  EXPECT_EQ(*(*e()), 10);
+}
+
+TEST(Task, ConstRefSuccess) {
+  int x = 10;
+  auto e = [&]() -> Task::Of<const int&> {
+    return Task::Success(std::cref(x));
+  };
+
+  auto e1 = [&]() {
+    return e()
+        | Then([](const int& value) {
+             return value + 10;
+           });
+  };
+
+  auto [future, k] = Terminate(e());
+
+  k.Start();
+
+  x = 42;
+
+  EXPECT_EQ(42, future.get());
+  EXPECT_EQ(52, *e1());
+}
+
+TEST(Task, ConstRefFunction) {
+  int x = 10;
+  auto e = [&]() -> Task::Of<const int&> {
+    return [&]() {
+      return Just(std::ref(x));
+    };
+  };
+
+  auto [future, k] = Terminate(e());
+
+  k.Start();
+
+  x = 42;
+
+  EXPECT_EQ(42, future.get());
+}
```

Comment:
Please add a newline at the end of this file.

Reply:
This should definitely be tested by our check_code_style workflow! Thanks @CodingCanuck!